emscripten-pty.js

```diff
@@ -89,20 +89,28 @@ Object.assign(Lib, {
         if (PTY_pollTimeout === 0) {
             return callback(PTY.readable);
         }
-        let handler, timeoutId;
+        let handlerReadable, handlerSignal, timeoutId;
         new Promise((resolve) => {
-            handler = PTY.onReadable(() => resolve(true));
+            handlerReadable = PTY.onReadable(() => resolve(0 /* ready */));
+
+            // We need to stop select(2) when a signal is caught.
+            //
+            // TODO: In fact, it should be stopped "when a signal handler is called," not "a the signal is received."
+            // If the signal handler is set as SIG_IGN, select(2) should not be stopped.
```

Comment:
You should be able to use `sigaction` to check it.

Reply:
Can I use `sigaction` from the JS side? Could you give me any pointers or examples?

Reply:
Sure, look at how `raise` is used in the same file already.

You can use any C function by adding like `'sigaction' ` to the relevant `__deps` array and calling it as `_sigaction` (note the unfortunate historical difference in underscore prefix).

Reply:
Thanks, I got it working with a patch as follows.

```diff
diff --git a/emscripten-pty.js b/emscripten-pty.js
index d81d51a..c02624c 100644
--- a/emscripten-pty.js
+++ b/emscripten-pty.js
@@ -84,7 +84,7 @@ Object.assign(Lib, {
         throw new FS.ErrnoError({{{ 1000 + cDefs.EAGAIN }}});
     },

-    $PTY_waitForReadableWithCallback__deps: ['$PTY', '$PTY_pollTimeout'],
+    $PTY_waitForReadableWithCallback__deps: ['malloc', 'free', 'sigaction', '$PTY', '$PTY_pollTimeout'],
     $PTY_waitForReadableWithCallback: (callback) => {
         if (PTY_pollTimeout === 0) {
             return callback(PTY.readable);
@@ -97,7 +97,16 @@ Object.assign(Lib, {
             //
             // TODO: In fact, it should be stopped "when a signal handler is called," not "a the signal is received."
             // If the signal handler is set as SIG_IGN, select(2) should not be stopped.
-            handlerSignal = PTY.onSignal(() => resolve(1 /* interrupted */));
+            handlerSignal = PTY.onSignal((signalName) => {
+                let signalCode = PTY_signalNameToCode[signalName];
+                var buff = _malloc({{{ POINTER_SIZE * 5 }}}); /* sizeof(struct sigaction) */
+                _sigaction(signalCode, null, buff);
+                var sa_handler = {{{ makeGetValue('buff', 0 /* sa_handler offset */, 'i32') }}};
+                _free(buff);
+                if (sa_handler != -2 /* SIG_IGN */) {
+                    resolve(1 /* interrupted */);
+                }
+            });

             if (PTY_pollTimeout >= 0) {
                 // if negative timeout, don't stop early (in poll-like functions it means infinite wait),
```

However, it was full of immediate values due to the lack of relevant struct and macro constant definitions.
Also, it should be handled for all signals, not just those generated by PTY.

Considering these things, I think that this support should be done on the emscripten side, not on xterm-pty. So I'd like to postpone this ad-hoc patch until it is truly needed.